### PR TITLE
AsyncGrpcClient: move class template to Call method

### PR DESCRIPTION
**Pull request by lebdron**
### Description of the Change
Move class template to Call method to allow reusing the same client for different endpoints.

### Benefits
Simpler code

### Possible Drawbacks 
None

### CODE REVIEW

**Comment by lebdron** [comment]
Ping @iceseer @kuvaldini

File: irohad/network/impl/async_grpc_client.hpp [hunk 1]
```diff
@@ -39,9 +38,7 @@ namespace iroha {
           if (not call->status.ok()) {
             log_->warn("RPC failed: {}", call->status.error_message());
           }
-          if (call->on_response) {
-            call->on_response(call->status, call->reply);
-          }
+          call->onResponse();
```
Comment (iceseer) [inline irohad/network/impl/async_grpc_client.hpp:41]: It seems a good location to catch exceptions and write into log file. Or even proceed to work.

Comment (iceseer) [reply irohad/network/impl/async_grpc_client.hpp:41]:
```c++
std::unique_ptr<AsyncClientCall> call(static_cast<AsyncClientCall*>(got_tag));
// delete call;

**Review by iceseer — CHANGES_REQUESTED** [review]

File: irohad/network/impl/async_grpc_client.hpp (hunk 1 above)
Comment (lebdron) [reply irohad/network/impl/async_grpc_client.hpp]: Fixed

**Comment by lebdron** [comment]
Ping @iceseer @kuvaldini

**Comment by lebdron** [comment]
Ping @iceseer @kuvaldini

**Review by iceseer — APPROVED** [review]

File: irohad/network/impl/async_grpc_client.hpp [hunk 2]
```diff
@@ -56,32 +59,44 @@ namespace iroha {
...
        */
-      template <typename F>
-      void Call(
-          F &&lambda,
-          std::function<void(grpc::Status &, Response &)> on_response = {}) {
-        auto call = new AsyncClientCall;
+      template <typename F, typename Response>
+      void Call(F &&lambda,
+                std::function<void(grpc::Status &, Response &)> on_response) {
+        auto call = new AsyncClientCallImpl<Response>;
```
Comment (kuvaldini) [inline irohad/network/impl/async_grpc_client.hpp:99]: Please use unique_ptr - do not forget to remove it

File: irohad/network/impl/async_grpc_client.hpp [hunk 3]
```diff
@@ -35,14 +34,18 @@ namespace iroha {
         void *got_tag;
         auto ok = false;
         while (cq_.Next(&got_tag, &ok)) {
-          auto call = static_cast<AsyncClientCall *>(got_tag);
+          std::unique_ptr<AsyncClientCall> call(
```
Comment (kuvaldini) [inline irohad/network/impl/async_grpc_client.hpp:37]: I see call created bellow with `new` is removed here. This is reinterpret_cast, not static and it looks bad.

Isn't there a way to keey strict types?

File: irohad/network/impl/async_grpc_client.hpp [hunk 4]
```diff
@@ -56,32 +59,44 @@ namespace iroha {
...
+      template <typename Response>
+      struct AsyncClientCallImpl : AsyncClientCall {
+        Response reply;
+
         std::unique_ptr<grpc::ClientAsyncResponseReaderInterface<Response>>
             response_reader;
 
         std::function<void(grpc::Status &, Response &)> on_response;
+
+        void onResponse() override {
```
Comment (kuvaldini) [inline irohad/network/impl/async_grpc_client.hpp:84]: This is called in the only place. Let's keep it simpler. Do actions inplace.

File: irohad/network/impl/async_grpc_client.hpp [hunk 5]
```diff
@@ -56,32 +59,44 @@ namespace iroha {
...
+        virtual void onResponse() = 0;
+
+        virtual ~AsyncClientCall() = default;
+      };
+
+      /**
+       * State and data information of gRPC call
+       */
+      template <typename Response>
+      struct AsyncClientCallImpl : AsyncClientCall {
```
Comment (kuvaldini) [inline irohad/network/impl/async_grpc_client.hpp:76]: Is there any other implementation?

If not - remove interface - it is useless

**Review by kuvaldini — CHANGES_REQUESTED** [review]

**Pull request closed without merging**